src/acl.c

```diff
@@ -714,6 +714,94 @@ void ACLRecomputeCommandBitsFromCommandRulesAllUsers(void) {
     raxStop(&ri);
 }
 
+static struct serverCommand *ACLLookupCommand(const char *name) {
+    struct serverCommand *cmd;
+    sds sdsname = sdsnew(name);
+    cmd = lookupCommandBySdsLogic(server.orig_commands, sdsname);
+    sdsfree(sdsname);
+    return cmd;
+}
+
+/* Remove stale module command rules from all users' selectors command_rules strings.
+ * This cleans up rules referring to commands/subcommands that no longer exist
+ * (e.g., after unloading a module), while preserving ACL categories. */
+void ACLCleanupStaleCommandRulesAllUsers(void) {
+    raxIterator ri;
+    raxStart(&ri,Users);
+    raxSeek(&ri,"^",NULL,0);
+    while(raxNext(&ri)) {
+        user *u = ri.data;
+        listIter li;
+        listNode *ln;
+        listRewind(u->selectors,&li);
+        while((ln = listNext(&li))) {
+            aclSelector *selector = (aclSelector *) listNodeValue(ln);
+            int argc = 0;
+            sds *argv = sdssplitargs(selector->command_rules, &argc);
+            if (!argv) continue;
+            sds new_rules = sdsempty();
+            for (int i = 0; i < argc; i++) {
+                sds tok = argv[i];
+                if (sdslen(tok) == 0) continue;
+                char first = tok[0];
+                if (first != '+' && first != '-') {
+                    /* Unknown formatting, keep as-is */
+                    new_rules = sdscatfmt(new_rules, "%S ", tok);
+                    continue;
+                }
+                /* Body after +/- */
+                const char *body = tok+1;
+                if (*body == '@') {
+                    /* Category: always keep */
+                    new_rules = sdscatfmt(new_rules, "%S ", tok);
+                    continue;
+                }
+                /* Command or subcommand. Extract main and optional sub after '|' */
+                const char *pipe = strchr(body, '|');
+                sds main = pipe ? sdsnewlen(body, pipe - body) : sdsnew(body);
+                struct serverCommand *cmd = ACLLookupCommand(main);
+                int keep = 0;
+                if (cmd) {
+                    if (!pipe) {
+                        keep = 1;
+                    } else {
+                        const char *sub = pipe + 1;
+                        if (cmd->subcommands_ht) {
+                            sds sub_sds = sdsnew(sub);
+                            struct serverCommand *sc = (struct serverCommand *)dictFetchValue(cmd->subcommands_ht, sub_sds);
```

Comment:
The command table uses `hashtable` instead of `dict` in 8.1 and later. That seems to be why the build is failing.

We probably need this code with dict when we backport the fix to 8.0 and 7.2 though.

Reply:
... or would it work to use `lookupCommandBySdsLogic` here?

UNTESTED:

```suggestion
                            struct serverCommand *sc = lookupCommandBySdsLogic(cmd->subcommands_ht, sub_sds);
```

Reply:
ah okay! thank you. will try and see if that helps